format/luajit/luajit.go

```diff
@@ -0,0 +1,381 @@
+
+import (
+	"bytes"
+	"embed"
+	"encoding/binary"
+
+	"github.com/wader/fq/format"
+	"github.com/wader/fq/pkg/decode"
+	"github.com/wader/fq/pkg/interp"
+	"github.com/wader/fq/pkg/scalar"
+)
+
+//go:embed luajit.md
+var LuaJITFS embed.FS
+
+var probeGroup decode.Group
+
+func init() {
+	interp.RegisterFormat(
+		format.LuaJIT,
+		&decode.Format{
+			Description: "LuaJIT 2.0 bytecode",
+			Groups:      []*decode.Group{format.Probe},
+			DecodeFn:    LuaJITDecode,
+			Dependencies: []decode.Dependency{
+				{Groups: []*decode.Group{format.Probe}, Out: &probeGroup},
+			},
+		})
+	interp.RegisterFS(LuaJITFS)
+}
+
+// reinterpret an int as a float
+func u64tof64(u uint64) float64 {
+	var buf [8]byte
+
+	binary.BigEndian.PutUint64(buf[:], u)
+
+	var f float64
+	err := binary.Read(bytes.NewBuffer(buf[:]), binary.BigEndian, &f)
+	if err != nil {
+		panic(err)
+	}
+
+	return f
+}
+
+type DumpInfo struct {
+	Strip     bool
+	BigEndian bool
+}
+
+func LuaJITDecodeHeader(di *DumpInfo, d *decode.D) {
+	d.FieldRawLen("magic", 3*8, d.AssertBitBuf([]byte{0x1b, 0x4c, 0x4a})) // ESC 'L' 'J'
+
+	d.FieldU8("version")
+
+	var flags uint64
+	d.FieldStruct("flags", func(d *decode.D) {
+		flags = d.FieldULEB128("raw")
+
+		d.FieldValueBool("be", flags&0x01 > 0)
+		d.FieldValueBool("strip", flags&0x02 > 0)
+		d.FieldValueBool("ffi", flags&0x04 > 0)
+		d.FieldValueBool("fr2", flags&0x08 > 0)
+	})
+
+	di.Strip = flags&0x2 > 0
+	di.BigEndian = flags&0x1 > 0
+
+	if !di.Strip {
+		namelen := d.FieldU8("namelen")
+		d.FieldUTF8("name", int(namelen))
+	}
+}
+
+type jumpBias struct{}
+
+func (j *jumpBias) MapUint(u scalar.Uint) (scalar.Uint, error) {
+	u.Actual -= 0x8000
+	return u, nil
+}
+
+func LuaJITDecodeBCIns(d *decode.D) {
+	op := d.FieldU8("op", opcodes)
+
+	d.FieldU8("a")
+
+	if opcodes[int(op)].HasD() {
+		if opcodes[int(op)].IsJump() {
+			d.FieldU16("j", &jumpBias{})
+		} else {
+			d.FieldU16("d")
+		}
+	} else {
+		d.FieldU8("c")
+		d.FieldU8("b")
+	}
+}
+
+type ktabType struct{}
+
+func (t *ktabType) MapUint(u scalar.Uint) (scalar.Uint, error) {
+	switch u.Actual {
+	case 0:
+		u.Sym = "nil"
+	case 1:
+		u.Sym = "false"
+	case 2:
+		u.Sym = "true"
+	case 3:
+		u.Sym = "int"
+	case 4:
+		u.Sym = "num"
+	default:
+		u.Sym = "str"
+	}
+	return u, nil
+}
+
+func LuaJITDecodeNum(d *decode.D) {
+	d.FieldAnyFn("num", func(d *decode.D) any {
+		lo := d.ULEB128()
+		hi := d.ULEB128()
+		return u64tof64((hi << 32) + lo)
+	})
+}
+
+func LuaJITDecodeKTabK(d *decode.D) {
+	ktabtype := d.FieldULEB128("ktabtype", &ktabType{})
+
+	if ktabtype >= 5 {
+		sz := ktabtype - 5
+		d.FieldUTF8("str", int(sz))
+	} else {
+		switch ktabtype {
+		case 3:
+			d.FieldULEB128("int")
+
+		case 4:
+			LuaJITDecodeNum(d)
+		}
+	}
+}
+
+func LuaJITDecodeCplx(d *decode.D) any {
+	lo := d.ULEB128()
+	if lo&1 == 0 {
+		return lo >> 1
+	} else {
+		hi := d.ULEB128()
+		return u64tof64((hi << 32) + (lo >> 1))
+	}
+}
+
+type kgcType struct{}
```

Comment:
One varint is to do something like:
```go
type fallbackUintMapSymStr struct {
	fallback string
	scalar.UintMapSymStr
}

func (m fallbackUintMapSymStr) MapUint(s scalar.Uint) (scalar.Uint, error) {
	s.Sym = m.fallback
	return m.UintMapSymStr.MapUint(s)
}
```
and then
```go
	kgctype := d.FieldULEB128("type", fallbackUintMapSymStr{
		fallback: "str",
		UintMapSymStr: scalar.UintMapSymStr{
			0: "child",
			1: "tab",
			2: "i64",
			3: "u64",
			4: "complex",
		},
	})
```
and use it for `ktabType` also